### Startup Success Prediction ###


**About Dataset**

Context:
A startup or start-up is a company or project begun by an entrepreneur to seek, develop, and validate a scalable economic model. While entrepreneurship refers to all new businesses, including self-employment and businesses that never intend to become registered, startups refer to new businesses that intend to grow large beyond the solo founder. 

Startups face high uncertainty and have high rates of failure, but a minority of them do go on to be successful and influential. 

Some startups become unicorns: privately held startup companies valued at over US$1 billion.

**The objective**

Predict whether a startup which is currently operating turns into a success or a failure. The success of a company is defined as the event that gives the company's founders a large sum of money through the process of M&A (Merger and Acquisition) or an IPO (Initial Public Offering). A company would be considered as failed if it had to be shut down.

**About the Data**

The data contains industry trends, investment insights and individual company information. There are 48 columns/features. Some of the features are:

- age_first_funding_year – quantitative
- age_last_funding_year – quantitative
- relationships – quantitative
- funding_rounds – quantitative
- funding_total_usd – quantitative
- milestones – quantitative
- age_first_milestone_year – quantitative
- age_last_milestone_year – quantitative
- state – categorical
- industry_type – categorical
- has_VC – categorical
- has_angel – categorical
- has_roundA – categorical
- has_roundB – categorical
- has_roundC – categorical
- has_roundD – categorical
- avg_participants – quantitative
- is_top500 – categorical
- status(acquired/closed) – categorical (the target variable, if a startup is ‘acquired’ by some other organization, means the startup succeed) 

These variables are likely binary indicators (0 or 1) provide information about the startup's funding history, specifically whether it has received funding from angel investors and whether it has successfully completed funding rounds at different stages (A, B, C, D). They can be valuable features in predicting the success or failure of a startup, as funding rounds and investment from angels are often associated with the growth and potential success of a company.
Notes: Each funding round represents a different stage of investment, and startups may not follow a strict sequential order (A, B, C, D). The order can vary based on the business's specific circumstances and the preferences of investors.

#### 1. Import the necessary libraires

In [1741]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

##### 2. Loading the database

In [1742]:
#Make sure you're aware of the current working directory in your Jupyter notebook.
import os
print(os.getcwd())

/Users/gabrielaarzate/Desktop/predicting_startup_succes/notebook


In [1743]:
# Assuming this notebook is in the 'notebook' directory
notebook_directory = '/Users/gabrielaarzate/Desktop/predicting_startup_succes/notebook'

In [1744]:
# Construct the full file path from the notebook directory
file_path = os.path.join(notebook_directory, '..', 'data', 'startup.csv')

#### 2. Read the data

In [1745]:
# It's time to read in our training and testing data using pd.read_csv, 
#and take a first look using the describe() function.
data = pd.read_csv(file_path, encoding="ISO-8859-1")

##### 3. Explore the data :

In [1746]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 923 entries, 0 to 922
Data columns (total 49 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                923 non-null    int64  
 1   state_code                923 non-null    object 
 2   latitude                  923 non-null    float64
 3   longitude                 923 non-null    float64
 4   zip_code                  923 non-null    object 
 5   id                        923 non-null    object 
 6   city                      923 non-null    object 
 7   Unnamed: 6                430 non-null    object 
 8   name                      923 non-null    object 
 9   labels                    923 non-null    int64  
 10  founded_at                923 non-null    object 
 11  closed_at                 335 non-null    object 
 12  first_funding_at          923 non-null    object 
 13  last_funding_at           923 non-null    object 
 14  age_first_

#### Target Variable ####

In [1747]:
data['status'].value_counts()
#In a binary classification problem like predicting whether a startup 
#will be "acquired" or "closed," a balanced dataset typically means that you have a roughly equal 
#number of examples for each class. In your case, you have 597 examples of "acquired" and 326 examples of "closed."

status
acquired    597
closed      326
Name: count, dtype: int64

##### 3.3 Analyse missing values 

In [1748]:
data.isnull().sum()

Unnamed: 0                    0
state_code                    0
latitude                      0
longitude                     0
zip_code                      0
id                            0
city                          0
Unnamed: 6                  493
name                          0
labels                        0
founded_at                    0
closed_at                   588
first_funding_at              0
last_funding_at               0
age_first_funding_year        0
age_last_funding_year         0
age_first_milestone_year    152
age_last_milestone_year     152
relationships                 0
funding_rounds                0
funding_total_usd             0
milestones                    0
state_code.1                  1
is_CA                         0
is_NY                         0
is_MA                         0
is_TX                         0
is_otherstate                 0
category_code                 0
is_software                   0
is_web                        0
is_mobil

In [1749]:
# Assuming 'data' is your DataFrame
null = pd.DataFrame(data.isnull().sum(), columns=["Null Values"])
null["% Missing Values"] = (data.isna().sum() / len(data) * 100)
null = null[null["% Missing Values"] > 0]

In [1750]:
styled_null = (
    null.style
    .background_gradient(cmap='viridis', low=0.2, high=0.1)
)

In [1751]:
styled_null
#We have 5 columns with missing values 

##### 3.4 Irrelevant features: 



- **Unnamed_0**: Appears to be an artifact of the data and often serves as an index or a row identifier.
- **id**: irrelevant 
- **Unnamed_6** : Repeating values from others columns(city, state_code,zip_code).
- **closed_at** : Irrelevant for now
- **state_code.1** : Has the same information as state.code column 
- **object_id** : Same information as id.
- **is_CA,is_NY,is_MA,is_TX,is_otherstate**: same information as state code.
- **is_software,is_web,is_mobile, is_enterprise, is_advertising, is_gamesvideo, is_ecommerce, is_biotech, is_consulting, is_othercategory** : same information as category_code.

We will drop 21 features that are irrelevant or repeating information.

#### 4 . Data Cleaning ####

4. 1 Drop the irrelevant features 

In [1752]:
data = data.drop(['Unnamed: 0', 'id', 'Unnamed: 6', 'closed_at', 'state_code.1', 'is_CA', 'is_NY', 'is_MA', 'is_TX', 'is_otherstate', 'is_software', 'is_web', 'is_mobile', 'is_enterprise', 'is_advertising', 'is_gamesvideo', 'is_ecommerce', 'is_biotech', 'is_consulting', 'is_othercategory', 'object_id'], axis=1).copy()
data

state_code   latitude   longitude zip_code           city  \
0           CA  42.358880  -71.056820    92101      San Diego   
1           CA  37.238916 -121.973718    95032      Los Gatos   
2           CA  32.901049 -117.192656    92121      San Diego   
3           CA  37.320309 -122.050040    95014      Cupertino   
4           CA  37.779281 -122.419236    94105  San Francisco   
..         ...        ...         ...      ...            ...   
918         CA  37.740594 -122.376471    94107  San Francisco   
919         MA  42.504817  -71.195611     1803     Burlington   
920         CA  37.408261 -122.015920    94089      Sunnyvale   
921         CA  37.556732 -122.288378    94404  San Francisco   
922         CA  37.386778 -121.966277    95054    Santa Clara   

                     name  labels founded_at first_funding_at last_funding_at  \
0             Bandsintown       1   1/1/2007         4/1/2009        1/1/2010   
1               TriCipher       1   1/1/2000        2/14/2005      12/28/2009   
2                   Plixi       1  3/18/2009        3/30/2010       3/30/2010   
3       Solidcore Systems       1   1/1/2002        2/17/2005       4/25/2007   
4          Inhale Digital       0   8/1/2010         8/1/2010        4/1/2012   
..                    ...     ...        ...              ...             ...   
918               CoTweet       1   1/1/2009         7/9/2009        7/9/2009   
919    Reef Point Systems       0   1/1/1998         4/1/2005       3/23/2007   
920       Paracor Medical       0   1/1/1999        6/29/2007       6/29/2007   
921               Causata       1   1/1/2009        10/5/2009       11/1/2011   
922  Asempra Technologies       1   1/1/2003        2/13/2006       2/13/2006   

     ...  category_code  has_VC  has_angel  has_roundA  has_roundB  \
0    ...          music       0          1           0           0   
1    ...     enterprise       1          0           0           1   
2    ...            web       0          0           1           0   
3    ...       software       0          0           0           1   
4    ...    games_video       1          1           0           0   
..   ...            ...     ...        ...         ...         ...   
918  ...    advertising       0          0           1           0   
919  ...       security       1          0           0           1   
920  ...        biotech       0          0           0           0   
921  ...       software       0          0           1           1   
922  ...       security       0          0           0           1   

     has_roundC  has_roundD  avg_participants is_top500    status  
0             0           0            1.0000         0  acquired  
1             1           1            4.7500         1  acquired  
2             0           0            4.0000         1  acquired  
3             1           1            3.3333         1  acquired  
4             0           0            1.0000         1    closed  
..          ...         ...               ...       ...       ...  
918           0           0            6.0000         1  acquired  
919           0           0            2.6667         1    closed  
920           0           1            8.0000         1    closed  
921           0           0            1.0000         1  acquired  
922           0           0            3.0000         1  acquired  

[923 rows x 28 columns]

In [1753]:
num_columns = len(data.columns)
print(f"After dropping irrelevant features The DataFrame has {num_columns} columns.")

After dropping irrelevant features The DataFrame has 28 columns.


4. 2 Handle Missing Values of 2 variables 

In [1754]:
#Before we had 5 columns with missing values, 3 of them were irrelevant(Unnamed: 6, state_code.1, closed_at) so 
#we dropped them now we need to input data to only 2 differents variables.

# 1. age_first_milestone_year
# 2. age_last_milestone_year 

# 2. Impute missing values with 0 to 'age_first_milestone_year'
data['age_first_milestone_year'].fillna(0, inplace=True)

# 3. Impute missing values with 0 to 'age_last_milestone_year'
data['age_last_milestone_year'].fillna(0, inplace=True)

In [1755]:
data.isnull().sum()

state_code                  0
latitude                    0
longitude                   0
zip_code                    0
city                        0
name                        0
labels                      0
founded_at                  0
first_funding_at            0
last_funding_at             0
age_first_funding_year      0
age_last_funding_year       0
age_first_milestone_year    0
age_last_milestone_year     0
relationships               0
funding_rounds              0
funding_total_usd           0
milestones                  0
category_code               0
has_VC                      0
has_angel                   0
has_roundA                  0
has_roundB                  0
has_roundC                  0
has_roundD                  0
avg_participants            0
is_top500                   0
status                      0
dtype: int64

In [1756]:
print(data.dtypes)

state_code                   object
latitude                    float64
longitude                   float64
zip_code                     object
city                         object
name                         object
labels                        int64
founded_at                   object
first_funding_at             object
last_funding_at              object
age_first_funding_year      float64
age_last_funding_year       float64
age_first_milestone_year    float64
age_last_milestone_year     float64
relationships                 int64
funding_rounds                int64
funding_total_usd             int64
milestones                    int64
category_code                object
has_VC                        int64
has_angel                     int64
has_roundA                    int64
has_roundB                    int64
has_roundC                    int64
has_roundD                    int64
avg_participants            float64
is_top500                     int64
status                      

4.3 Analyzing Categorical Data to convert to Numerical Data 

In [1757]:
numerical_features = data.select_dtypes(include=['number']).columns.tolist()
categorical_features = data.select_dtypes(include=['object']).columns.tolist()

# Assuming the target variable is 'status'
target_variable = ['status']

# Print the lists along with the number of features
print("Numerical Features ({0}):".format(len(numerical_features)))
print(numerical_features)

print("\nCategorical Features ({0}):".format(len(categorical_features)))
print(categorical_features)

print("\nTarget Variable ({0}):".format(len(target_variable)))
print(target_variable)

Numerical Features (19):
['latitude', 'longitude', 'labels', 'age_first_funding_year', 'age_last_funding_year', 'age_first_milestone_year', 'age_last_milestone_year', 'relationships', 'funding_rounds', 'funding_total_usd', 'milestones', 'has_VC', 'has_angel', 'has_roundA', 'has_roundB', 'has_roundC', 'has_roundD', 'avg_participants', 'is_top500']

Categorical Features (9):
['state_code', 'zip_code', 'city', 'name', 'founded_at', 'first_funding_at', 'last_funding_at', 'category_code', 'status']

Target Variable (1):
['status']


In [1758]:
#1. Categorical Data ir Nominal or Ordinal?

# Assuming your_data_df is your main DataFrame
your_data_df = data

# List of your categorical columns
categorical_columns =  your_data_df.select_dtypes(include=['object']).columns

# Loop through each categorical column and display a subset of unique values
for categorical_column in categorical_columns:
    unique_values = your_data_df[categorical_column].unique()
    
    print(f"\nColumn: {categorical_column}")
    print("Unique Values (Subset):")
    print(unique_values[:10]) 


Column: state_code
Unique Values (Subset):
['CA' 'MA' 'KY' 'NY' 'CO' 'VA' 'TX' 'WA' 'IL' 'NC']

Column: zip_code
Unique Values (Subset):
['92101' '95032' '92121' '95014' '94105' '94043' '94041' '94901' '1267'
 '94306']

Column: city
Unique Values (Subset):
['San Diego' 'Los Gatos' 'Cupertino' 'San Francisco' 'Mountain View'
 'San Rafael' 'Williamstown' 'Palo Alto' 'Menlo Park' 'Louisville']

Column: name
Unique Values (Subset):
['Bandsintown' 'TriCipher' 'Plixi' 'Solidcore Systems' 'Inhale Digital'
 'Matisse Networks' 'RingCube Technologies' 'ClairMail' 'VoodooVox'
 'Doostang']

Column: founded_at
Unique Values (Subset):
['1/1/2007' '1/1/2000' '3/18/2009' '1/1/2002' '8/1/2010' '1/1/2005'
 '1/1/2004' '6/1/2005' '11/15/2000' '1/1/2006']

Column: first_funding_at
Unique Values (Subset):
['4/1/2009' '2/14/2005' '3/30/2010' '2/17/2005' '8/1/2010' '7/18/2006'
 '9/21/2006' '8/24/2005' '8/2/2005' '2/1/2007']

Column: last_funding_at
Unique Values (Subset):
['1/1/2010' '12/28/2009' '3/30/2010'

4.4 Converting Datetime to Numerical int64

- these variables : 'founded_at', 'first_funding_at', 'last_funding_at'

In [1759]:
import pandas as pd

# Assuming your_data_df is your main DataFrame
your_data_df = data

# List of date columns
date_columns = ['founded_at', 'first_funding_at', 'last_funding_at']

# Convert date columns to datetime
for date_column in date_columns:
    your_data_df[date_column] = pd.to_datetime(your_data_df[date_column], errors='coerce')

# Convert datetime to timestamp (numerical)
for date_column in date_columns:
    your_data_df[date_column] = your_data_df[date_column].astype(int)

# Display the result
print(your_data_df.dtypes)


state_code                   object
latitude                    float64
longitude                   float64
zip_code                     object
city                         object
name                         object
labels                        int64
founded_at                    int64
first_funding_at              int64
last_funding_at               int64
age_first_funding_year      float64
age_last_funding_year       float64
age_first_milestone_year    float64
age_last_milestone_year     float64
relationships                 int64
funding_rounds                int64
funding_total_usd             int64
milestones                    int64
category_code                object
has_VC                        int64
has_angel                     int64
has_roundA                    int64
has_roundB                    int64
has_roundC                    int64
has_roundD                    int64
avg_participants            float64
is_top500                     int64
status                      

Now variables in numerical columns not in categorical

In [1760]:

# Assuming your_data_df is your main DataFrame
your_data_df = data

# List of your categorical columns
numerical_columns = your_data_df.select_dtypes(include=['number']).columns

# Loop through each categorical column and display a subset of unique values
for numerical_columns in numerical_columns:
    unique_values = your_data_df[numerical_columns].unique()
    
    print(f"\nColumn: {numerical_columns}")
    print("Unique Values (Subset):")
    print(unique_values[:10]) 


Column: latitude
Unique Values (Subset):
[42.35888   37.238916  32.901049  37.320309  37.779281  37.406914
 37.3915589 38.057107  42.712207  37.427235 ]

Column: longitude
Unique Values (Subset):
[ -71.05682   -121.973718  -117.192656  -122.05004   -122.419236
 -122.09037   -122.0702643 -122.513742   -73.203599  -122.145783 ]

Column: labels
Unique Values (Subset):
[1 0]

Column: founded_at
Unique Values (Subset):
[1167609600000000000  946684800000000000 1237334400000000000
 1009843200000000000 1280620800000000000 1104537600000000000
 1072915200000000000 1117584000000000000  974246400000000000
 1136073600000000000]

Column: first_funding_at
Unique Values (Subset):
[1238544000000000000 1108339200000000000 1269907200000000000
 1108598400000000000 1280620800000000000 1153180800000000000
 1158796800000000000 1124841600000000000 1122940800000000000
 1170288000000000000]

Column: last_funding_at
Unique Values (Subset):
[1262304000000000000 1261958400000000000 1269907200000000000
 1177459200

In [1761]:
# Assuming your_data_df is your main DataFrame
your_data_df = data

# List of your categorical columns
categorical_columns =  your_data_df.select_dtypes(include=['object']).columns

# Loop through each categorical column and display a subset of unique values
for categorical_column in categorical_columns:
    unique_values = your_data_df[categorical_column].unique()
    
    print(f"\nColumn: {categorical_column}")
    print("Unique Values (Subset):")
    print(unique_values[:100]) 


Column: state_code
Unique Values (Subset):
['CA' 'MA' 'KY' 'NY' 'CO' 'VA' 'TX' 'WA' 'IL' 'NC' 'PA' 'GA' 'NH' 'MO'
 'FL' 'NJ' 'WV' 'MI' 'DC' 'CT' 'MD' 'OH' 'TN' 'MN' 'RI' 'OR' 'UT' 'ME'
 'NV' 'NM' 'IN' 'AZ' 'ID' 'AR' 'WI']

Column: zip_code
Unique Values (Subset):
['92101' '95032' '92121' '95014' '94105' '94043' '94041' '94901' '1267'
 '94306' '94025' '40204' '11201' '80202' '22182' '94022' '94010' '10004'
 '94301' '78735' '98122' '94103' '80302' '60601' '94303' '94704' '78701'
 '92705' '94035' '98119' '27701' '15219' '10011' '94111' '95134' '94107'
 '10010' '30303' '3101' '98102' '94089' '94085' '100011' '2139' '94607'
 '94404' '94080' '2210' '78731' '2451' '95128' '92656' '64106' '1887'
 '98033-6314' '33609' '02111-1720' '94501' '98011' '94403-1855' '94104'
 '94087' '33626' '75001' '94538' '95051' '2138' '8540' '80538' '10012'
 '25430' '90025' '48187' '98004' '95054' '20007' '94110' '10173' '60606'
 '8873' '94710' '30009' '98104' '22902' '20166' '60607' '6002' '60401'
 '95035-6261' '

Creating Dummies for categorical variables

In [1762]:
test_dummies = pd.get_dummies(data, drop_first=False)

In [1763]:
test_dummies.head()

latitude   longitude  labels           founded_at     first_funding_at  \
0  42.358880  -71.056820       1  1167609600000000000  1238544000000000000   
1  37.238916 -121.973718       1   946684800000000000  1108339200000000000   
2  32.901049 -117.192656       1  1237334400000000000  1269907200000000000   
3  37.320309 -122.050040       1  1009843200000000000  1108598400000000000   
4  37.779281 -122.419236       0  1280620800000000000  1280620800000000000   

       last_funding_at  age_first_funding_year  age_last_funding_year  \
0  1262304000000000000                  2.2493                 3.0027   
1  1261958400000000000                  5.1260                 9.9973   
2  1269907200000000000                  1.0329                 1.0329   
3  1177459200000000000                  3.1315                 5.3151   
4  1333238400000000000                  0.0000                 1.6685   

   age_first_milestone_year  age_last_milestone_year  ...  \
0                    4.6685                   6.7041  ...   
1                    7.0055                   7.0055  ...   
2                    1.4575                   2.2055  ...   
3                    6.0027                   6.0027  ...   
4                    0.0384                   0.0384  ...   

   category_code_security  category_code_semiconductor  category_code_social  \
0                   False                        False                 False   
1                   False                        False                 False   
2                   False                        False                 False   
3                   False                        False                 False   
4                   False                        False                 False   

   category_code_software  category_code_sports  category_code_transportation  \
0                   False                 False                         False   
1                   False                 False                         False   
2                   False                 False                         False   
3                    True                 False                         False   
4                   False                 False                         False   

   category_code_travel  category_code_web  status_acquired  status_closed  
0                 False              False             True          False  
1                 False              False             True          False  
2                 False               True             True          False  
3                 False              False             True          False  
4                 False              False            False           True  

[5 rows x 1619 columns]

Nominal Data = named data (No Order):example = gender(m/f), age, groups Use One-Hot Encoding.
Ordinal Data (Clear Order): example =  1,2,3, Education = High School, Middle Schoo, =  Use Label Encoding.

In our analysis, the majority of the categorical data exhibited a nominal nature, Consequently, we opted for one-hot encoding to represent each category as a binary column. This method ensures the independence of categories, eliminating the risk of introducing unintended ordinal relationships. 

Update: I have a problem because when using One hot enconding it creates binary features, and it lead to a large number 
of new features, so i need to find another way to convert the categorical data to numerical.
Update 2: #As I expected i was triying to us get_dummies(one-hot encoding ) but it creates 1619 columns way to long 
#create a new column for each unique value, so this is not an option.

#### Detecting outliers

#### 2 .  Data Analysiss ####

Now that our data is cleaned, we will explore our data with descriptive and graphical statistics to describe and summarize our variables. In this stage, you will find yourself classifying features and determining their correlation with the target variable and each other.

In [1764]:
# # Now use 'encoded_data' in the correlation calculation
# data_encoded = data_encoded
# column_names = data_encoded.columns.tolist()
# print(column_names)

In [1765]:

# # Now use 'encoded_data' in the correlation calculation
# data_encoded = data_encoded
# column_names = data_encoded.columns.tolist()

# # Identify the column name representing 'status' after one-hot encoding
# status_column_name = [col for col in column_names if 'status' in col.lower()][0]


# # Extracting feature matrix
# feature_matrix = your_data_df[column_names]

# def calculate_correlations(df: pd.DataFrame, target: pd.Series) -> pd.DataFrame:
#     """
#     Calculates the correlations of all columns with regards to the target and returns a DataFrame with all column names
#     and their correlation coefficient with regards to the target.

#     :param df: Pandas DataFrame
#     :param target: Pandas Series with target values
#     :return: Pandas DataFrame with column names and their correlation coefficient with regards to the target
#     """
#     # Calculate correlations
#     correlations = df.corrwith(target)
    
#     # Create a DataFrame with column names and correlation coefficients
#     correlation_df = pd.DataFrame(correlations, columns=['correlation_coefficient'])
    
#     return correlation_df

# # Now you can calculate correlations using the identified 'status' column name
# xcorrelations = calculate_correlations(feature_matrix, encoded_data[status_column_name])

# # Display the correlations
# print(xcorrelations)

#### 3. Preprocessing

Tranforming status(target variable) from categorical to numerical values 

2.Detect and address outliers.

3.Remove or impute incorrect or inconsistent data.

Statistical Summary

#### Define the inputs and the target

#### 5- Data Segmentation

#### 6- Model Training

#### 7- Model Testing

#### 8-  Model Evaluation:
#### Confusion Matrix and Analysis:

#### 10 - Print a classification report